backend/src/v2/driver/driver.go

```diff
@@ -2006,7 +2028,253 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
+	starterTaskQueue := make([]string, 0)
+	parentDAG, err := cfg.mlmd.GetExecution(cfg.ctx, currentTask.GetExecution().GetCustomProperties()["parent_dag_id"].GetIntValue())
+	if err != nil {
+		return nil, err
+	}
+	resolveCollection, iterations := getParallelForIterationCount(currentTask, parentDAG)
+
+	glog.V(4).Infof("resolveCollection: %v", resolveCollection)
+	glog.V(4).Infof("iterations: %v", iterations)
+	if resolveCollection {
+		for i := range iterations {
+			starterTaskName := metadata.UpdateTaskNameWithDagID(currentTask.TaskName(), currentTask.GetID())
+			starterTaskName = metadata.GetParallelForTaskName(starterTaskName, i)
+			starterTaskQueue = append(starterTaskQueue, starterTaskName)
+		}
+	}
+	return starterTaskQueue, nil
+}
+
+// In order to properly perform artifact/parameter collection across a DAG/tree
+// we perform a breadth first search traversal.
+func CollectInputs(cfg resolveUpstreamOutputsConfig, starterTaskQueue []string, tasks map[string]*metadata.Execution, outputKey string, isArtifact bool) (*structpb.Value, *pipelinespec.ArtifactList, error) {
+	glog.V(4).Infof("currentTask is a ParallelFor DAG. Attempting to gather all nested producer_subtasks")
+	// Set some helpers for the BFS
+	parallelForParameterList := make([]*structpb.Value, 0)
+	parallelForArtifactList := make([]*pipelinespec.RuntimeArtifact, 0)
+	resolveTaskNameQueue := starterTaskQueue
+	var currentTask *metadata.Execution
+	var workingSubTaskName string
+	workingOutputKey := outputKey
+	tempWorkingOutputKey := outputKey
+	previousTaskName := tasks[resolveTaskNameQueue[0]].TaskName()
+
+	for key, task := range tasks {
+		glog.Infof("task: %v  || task: %v", key, task)
+	}
+	for len(resolveTaskNameQueue) > 0 {
+		// The starterQueue contains the first set of child DAGs from the
+		// parallelFor, i.e. the iteration dags
+		glog.V(4).Infof("resolveTaskNameQueue: %v", resolveTaskNameQueue)
+		currentTaskName := resolveTaskNameQueue[0]
+		resolveTaskNameQueue = resolveTaskNameQueue[1:]
+
+		currentTask = tasks[currentTaskName]
+
+		// We check if these values need to be updated going through the
+		// resolution of dags/tasks Most commonly the subTaskName will change
+		// for both parameter & artifact resolution. For parameter resolutions,
+		// the outputParameterKey can change, and is used for extracting the
+		// appropriate field off of the struct set for the outputs on the task
+		// in question.
+
+		// An issue arises if we update the outputParameterKey but there exists
+		// multiple iterations of the same task and we haven't fully parsed all
+		// iterations. We will encounter a scenario where we will attempt to
+		// extract fields from the struct with the wrong key. Hence, the
+		// condition below. NOTE: This is only an issue for Parameter resolution
+		// and does not interfere with Artifact resolution
+		if currentTask.TaskName() == previousTaskName {
+			workingOutputKey = tempWorkingOutputKey
```

Comment:
Should `tempWorkingOutputKey` just be called `previousWorkingOutputKey` or something?